ckan/templates-midnight-blue/header.html

```diff
@@ -1,125 +1,119 @@
+              <a class="logo" href="{{ h.url_for('home.index') }}">
+                <img src="{{ h.url_for_static_or_external(g.site_logo) }}" alt="{{ g.site_title }}"
+                  aria-label="{{ _('Navigate to Home - {site_title}').format(site_title=g.site_title) }}"/>
+              </a>
+            {% else %}
+              <a href="{{ h.url_for('home.index') }}">{{ g.site_title }}</a>
+              {% if g.site_description %}
+                <p class="mb-0">{{ g.site_description }}</p>
+              {% endif %}
+            {% endif %}
+          {% endblock %}
+        </div>
+        <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#main-navigation-toggle"
+          aria-controls="main-navigation-toggle" aria-expanded="false" aria-label="{{ _('Toggle navigation') }}">
+          <span class="fa fa-bars"></span>
+        </button>
+        <div class="main-navbar collapse navbar-collapse col-md-10" id="main-navigation-toggle">
           {% block header_site_navigation %}
-            {% block header_site_navigation_tabs %}
-              {% set org_type = h.default_group_type('organization') %}
-              {% set group_type = h.default_group_type('group') %}
-
-              {{ h.build_nav_main(
-                (dataset_type ~ '.search', h.humanize_entity_type('package', dataset_type, 'main nav') or _('Datasets'), ["dataset", "resource"]),
-                (org_type ~ '.index',
-                  h.humanize_entity_type('organization', org_type, 'main nav') or _('Organizations'), ['organization']),
-                (group_type ~ '.index',
-                  h.humanize_entity_type('group', group_type, 'main nav') or _('Groups'), ['group']),
-                ('home.about', _('About')) ) }}
-            {% endblock %}
-          </ul>
-
-      {% endblock %}
-      {% block header_site_search %}
-          <form class="d-flex site-search" action="{% url_for dataset_type ~ '.search' %}" method="get">
-              <label class="d-none" for="field-sitewide-search">{% block header_site_search_label %}{{ h.humanize_entity_type('package', dataset_type, 'search_placeholder') or _('Search Datasets...') }}{% endblock %}</label>
-              <input id="field-sitewide-search" class="form-control me-2"  type="text" name="q" placeholder="{{ _('Search') }}" aria-label="{{ h.humanize_entity_type('package', dataset_type, 'search_placeholder') or _('Search datasets...') }}"/>
-              <button class="btn" type="submit" aria-label="{{ _('Submit') }}"><i class="fa fa-search"></i></button>
-          </form>
-      {% endblock %}
-      </div>
-    </nav>
-  </div>
-</header>
-{% endblock %}
+            <ul class="navbar-nav mb-2 mb-lg-0 col-md-10">
+              {% block header_site_navigation_tabs %}
+                {% set org_type = h.default_group_type('organization') %}
+                {% set group_type = h.default_group_type('group') %}
+                {{ h.build_nav_main(
```

Comment:
In general, for the new theme we want to try to move away from python helpers _that output HTML_ (not python helpers in general) in favour of straight Jinja code or Jinja macros. This makes it much easier for plugins to customize and avoids the spaghetti code we now have.

.e.g. `build_nav_main` calls `_make_menu_item` calls `_link_active` and `_link_to`, which calls `_link_class`, `_create_link_text` and `_link_aria_current` to finally call `link_to`. 

And all this to output:

```html
<li class="active"><a aria-current="page" href="/dataset/">Datasets</a></li>
<li><a href="/organization/">Organizations</a></li>
<li><a href="/group/">Groups</a></li>
<li><a href="/about">About</a></li>
```

The same could be achieved with something like (`h.page_is_active()` does not exist, we can copy whatever logic is used now):

```
<li {% if h.page_is_active() %}class="active{% endif %}>
  <a {% if h.page_is_active() %}aria-current="page"{% endif %} href="{% h.url_for(dataset_type ~ '.search') }">
  {% h.humanize_entity_type('package', dataset_type, 'main nav') or _('Datasets') %}
  </a>
</li>
...
```
If we wanted to make it slightly more convenient we can create a Jinja macro in `ckan/templates/macros/nav_link.html`:

```
{% macro nav_link(url, text): %}
<li {% if h.page_is_active() %}class="active{% endif %}>
  <a {% if h.page_is_active() %}aria-current="page"{% endif %} href="{{ url }}">
  {{ text }}
  </a>
</li>
```
and then call it from `header.html`:

```
{% import 'macros/nav_link.html' as nav_link %}

{{ nav_link(h.url_for(dataset_type ~ '.search'),  h.humanize_entity_type('package', dataset_type, 'main nav') or _('Datasets')) }}
{{ nav_link(h.url_for(org_type ~ '.index'),  h.humanize_entity_type('organization', org_type, 'main nav') or _('Organizations')) }}
{{ nav_link(h.url_for(group_type ~ '.index'),  h.humanize_entity_type('group', group_type, 'main nav') or _('Grpups')) }}
{{ nav_link(h.url_for('home.about'), _('About')) }}
```

@aleeexgreeen hope this makes sense, I'm keen to hear your thoughts on this, what form do you think is more convenient/user friendly?

Reply:
Hi! It makes a lot of sense to do it this way, it will simplify the process of adding attributes and elements to the links in the future. 

I think going with the macro is the best idea, I can see it being used for breadcrumbs, tabs and potentially pagination as well, which all require an `aria-current` attribute for better accessibility (since the tabs being used are links `aria-current="page"` is appropriate to use instead of `aria-selected="true"`). 

The macro solution is super convenient if we can pass additional classes, an icon, the url, and the link text (those are the only ones I can think of right now). It will be used in place of helpers `build_nav_main`, `build_nav_icon` and `build_nav` , making `_make_menu_item` obsolete, so I'm all for this option.